src/Text/Pandoc/FlavoredFormat.hs

```diff
@@ -0,0 +1,76 @@
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeApplications  #-}
+{- |
+   Module      : Text.Pandoc.FlavoredFormat
+   Copyright   : © 2022 Albert Krewinkel
+   License     : GPL-2.0-or-later
+   Maintainer  : Albert Krewinkel <tarleb+pandoc@moltkeplatz.de>
+
+Handling of format specifiers for input and output.
+-}
+module Text.Pandoc.FlavoredFormat
+  ( FlavoredFormat (..)
+  , parseFlavoredFormat
+  ) where
+
+import Control.Monad.Except (throwError)
+import Text.Pandoc.Class (PandocMonad)
+import Text.Pandoc.Error (PandocError (..))
+import Text.Pandoc.Extensions
+  ( Extension
+  , Extensions
+  , disableExtension
+  , enableExtension
+  , extensionEnabled
+  , getAllExtensions
+  , getDefaultExtensions
+  , parseFormatSpec
+  )
+import qualified Data.Text as T
+
+-- | Format specifier with the format's name and the lists of extensions
+-- to be enabled or disabled.
+data FlavoredFormat = FlavoredFormat
+  { formatName       :: T.Text
+  , formatExtensions :: Extensions
+  }
+
+-- | Changes to a set of extensions, i.e., list of extensions to be
+-- enabled or disabled.
+data ExtensionsDiff = ExtensionsDiff
+  { extsToEnable  :: [Extension]
+  , extsToDisable :: [Extension]
+  }
+
+-- | Parses a format with extensions; enables and/or disables the
+-- specified extensions, starting from the default extensions for the
+-- given base format. Errors are thrown if the format cannot be parsed,
+-- or if the given extensions are not supported for the format.
+parseFlavoredFormat :: PandocMonad m
+                    => T.Text -> m FlavoredFormat
+parseFlavoredFormat fmtTxt =
+  case parseFormatSpec fmtTxt of
+    Left err -> throwError $ PandocFormatError fmtTxt (T.pack $ show err)
+    Right (fname, extsToEnable', extsToDisable') -> do
+      let extsDiff = ExtensionsDiff extsToEnable' extsToDisable'
+      let allExts = getAllExtensions fname
+      let defExts = getDefaultExtensions fname
+      case getUnsupportedExtensions allExts extsDiff of
+        ext:_ -> throwError $ PandocUnsupportedExtensionError
+                 (T.drop 4 . T.pack $ show ext) fname
+        _ -> pure $ FlavoredFormat fname (applyExtensionsDiff defExts extsDiff)
+
+-- | Apply the lists of extensions to enable and disable to a set of
+-- extensions.
+applyExtensionsDiff :: Extensions -> ExtensionsDiff -> Extensions
+applyExtensionsDiff exts extsDiff =
+  foldr disableExtension
+        (foldr enableExtension exts (extsToEnable extsDiff))
+        (extsToDisable extsDiff)
+
+-- | Returns the list of unsupported extensions in the extensions diff.
+getUnsupportedExtensions :: Extensions -> ExtensionsDiff -> [Extension]
+getUnsupportedExtensions allExts extsDiff =
+  filter (\ext -> not $ extensionEnabled ext allExts)
+         (extsToEnable extsDiff ++ extsToDisable extsDiff)
```

Comment:
Instead of having two functions, would it make sense to have just one?
``` hs
applyExtensionsDiff :: Extensions -> ExtensionsDiff -> Either [Extension] Extensions
```

Reply:
The `Extensions` argument for `applyExtensionsDiff` will usually be the default extensions, and for `getUnsupportedExtensions` the set of all supported extensions. So `applyExtensionsDiff` would need an extra argument, and I always get confused about which is which if a function takes two arguments of the same type.

Reply:
There's now a new type `ExtensionsConfig` that holds the sets of default and supported extensions. `applyExtensionsDiff` was rewritten to use that instead. The new type will be handy for extensions support in custom writers, but it might need a better name.